File: frontend/src/api/node.ts
import { apiGet, apiPost, apiPut, apiPatch, apiDelete } from './http'

export interface NodeCreate {
  nodeTypeId: string
  name: string
  dn?: string | null
  status?: string
}

export interface NodeUpdate {
  name?: string | null
  dn?: string | null
  status?: string | null
}

export interface NodePositionUpdate {
  x: number
  y: number
}

export interface NodeDetail {
  id: string
  topologyId: string
  nodeTypeId: string
  name: string
  dn: string | null
  status: string
  createdAt: string
  updatedAt: string
  x: number | null
  y: number | null
  attrs?: Record<string, string | null>
}

export interface NodeItem {
  id: string
  topologyId: string
  nodeTypeId: string
  name: string
  dn: string | null
  status: string
  createdAt: string
  updatedAt: string
  x: number | null
  y: number | null
}

export interface NodeListResponse {
  items: NodeItem[]
  total: number
  page: number
  pageSize: number
}

export interface BulkNodeItem {
  name: string
  x: number
  y: number
  attrs: Record<string, string | null>
}

export interface BulkNodesCreateRequest {
  nodeTypeId: string
  items: BulkNodeItem[]
}

export interface BulkCreatedItem {
  index: number
  id: string
  name: string
}

export interface BulkSkippedItem {
  index: number
  name: string | null
  reason: string
}

export interface BulkNodesCreateResponse {
  created: BulkCreatedItem[]
  skipped: BulkSkippedItem[]
}

export const nodeApi = {
  list: (topologyId: string, params?: { nodeTypeId?: string; status?: string; page?: number; pageSize?: number }): Promise<NodeListResponse> =>
    apiGet(`/topologies/${topologyId}/nodes`, params),

  get: (id: string): Promise<NodeDetail> =>
    apiGet(`/nodes/${id}`),

  create: (topologyId: string, data: NodeCreate): Promise<{ id: string }> =>
    apiPost(`/topologies/${topologyId}/nodes`, data),

  update: (id: string, data: NodeUpdate): Promise<{ id: string }> =>
    apiPut(`/nodes/${id}`, data),

  delete: (id: string): Promise<{ id: string }> =>
    apiDelete(`/nodes/${id}`),

  updatePosition: (id: string, data: NodePositionUpdate): Promise<{ id: string }> =>
    apiPatch(`/nodes/${id}/position`, data),

  setAttrs: (id: string, attrs: Record<string, string | null>): Promise<{ id: string }> => {
    const attrsList = Object.entries(attrs).map(([field_key, value]) => ({ field_key, value }))
    return apiPut(`/nodes/${id}/attrs`, attrsList)
  },

  bulkCreate: (
    topologyId: string,
    data: BulkNodesCreateRequest,
  ): Promise<BulkNodesCreateResponse> =>
    apiPost(`/topologies/${topologyId}/nodes/bulk`, data),
}

File: backend/app/admin/schemas/node.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field

from ._base import CamelModel


class NodeCreate(CamelModel):
    node_type_id: str
    name: str = Field(..., min_length=1, max_length=100)
    dn: Optional[str] = Field(default=None, max_length=200)
    status: str = Field(default="online", max_length=20)


class NodeUpdate(CamelModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    dn: Optional[str] = Field(default=None, max_length=200)
    status: Optional[str] = Field(default=None, max_length=20)


class NodePositionUpdate(CamelModel):
    x: float
    y: float


class NodeAttrSet(CamelModel):
    field_key: str
    value: Optional[str] = None


class NodeItem(CamelModel):
    id: str
    topology_id: str
    node_type_id: str
    name: str
    dn: Optional[str]
    status: str
    created_at: datetime
    updated_at: datetime


class NodeListItem(NodeItem):
    x: Optional[float] = None
    y: Optional[float] = None


class NodeDetail(NodeItem):
    attrs: dict[str, Any] = {}
    x: Optional[float] = None
    y: Optional[float] = None


class NodeListResponse(CamelModel):
    code: int = 0
    data: dict[str, Any]
    message: str = "ok"


class BulkNodeItem(CamelModel):
    name: str = Field(..., min_length=1, max_length=100)
    x: float
    y: float
    attrs: dict[str, Optional[str]] = Field(default_factory=dict)


class BulkNodesCreateRequest(CamelModel):
    node_type_id: str
    items: list[BulkNodeItem]


class BulkCreatedItem(CamelModel):
    index: int
    id: str
    name: str


class BulkSkippedItem(CamelModel):
    index: int
    name: Optional[str] = None
    reason: str


class BulkNodesCreateResponse(CamelModel):
    created: list[BulkCreatedItem]
    skipped: list[BulkSkippedItem]
